ボウリングスコア計算プログラムを作成

sthnu32424 opened the pull request.

sthnu32424 committed:
feat: initial implementation of bowling score calculator

sthnu32424 committed:
refactor: bowling score calculation logic for clarity and efficiency

- Remove unnecessary variable `params`
- Simplify frame processing by removing `[10, 0]` cases
- Flatten frames after the 10th one
- Refactor strike and spare calculations for better readability
- Add `# frozen_string_literal: true` for performance improvement

sthnu32424 committed:
refactor: Improve code readability and simplify comments

- Updated comments for better clarity and consistency
- Renamed variables and comments to better reflect their purpose
- Simplified strike and spare calculation logic for improved readability

sthnu32424 committed:
test: add test cases for bowling score calculation

maedana commented on 03.bowling/test_bowling.rb:
```diff
@@ -0,0 +1,41 @@
+
+require 'minitest/autorun'
+
+class BowlingTest < Minitest::Test
```
テストコードも書いてあっていいですね :+1:

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+frames = []
+shots.each_slice(2) do |shot|
+  frames << shot
+end
```
`frames =shots.each_slice(2).to_a`とかけそうですね

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+
+# スコア計算
+point = 0
+point += frames.each_with_index.sum do |frame, index|
```
最初から`point = frames.each_with...`と書いてしまえばよさそうですね

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+# スコアをカンマ区切りで取得
```
いわゆる説明コメントにあたるのでなるべく書かないようにしたいですね。慣れないうちは無理に削らなくてもいいですが。

参考
https://qiita.com/jnchito/items/f0d90af4ed44b7484103

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+shots = []
+scores.each do |score|
+  if score == 'X'
+    shots << 10
```
ストライクの意味での10が繰り返し出てくるので定数化するなどしてリーダブルにしたいですね

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+
+  if frame[0] == 10 # ストライクのとき
+    # 次の2投を加算する（次のフレームがストライクなら、さらにその次の1投も考慮）
+    next_two_shots = frames[index + 1][1] || frames[index + 2][0]
```
これ次の2投のうちの2投目ですよね。命名が`next_two_shots`はおかしくないですか? この命名ですと、2投分のサイズ2の配列かなと思ってしまいます。

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+  if frame[0] == 10 # ストライクのとき
+    # 次の2投を加算する（次のフレームがストライクなら、さらにその次の1投も考慮）
+    next_two_shots = frames[index + 1][1] || frames[index + 2][0]
+    10 + frames[index + 1][0] + next_two_shots
```
動的値で表現出来るものは動的値で表現したほうがいいんじゃないですかね。この場合の`10`って固定でなく`frame`を使った動的値に出来ると思うんです。

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,47 @@
+    10 + frames[index + 1][0] + next_two_shots
+  elsif frame.sum == 10 # スペアのとき
+    # 次の1投を加算
+    10 + frames[index + 1][0]
```
ここの10も動的値にできそうですね

sthnu32424 committed:
refactor: Remove unnecessary code and correct variable names.

- Simplify the process of creating frames.
- Direct assignment to point.
- Change next_to_shots to second_bonus_ball.
- Change 10 to frame.sum

sthnu32424 committed:
refactor: Delete unnecessary comments.

Because it was an explanatory comment.

sthnu32424 committed:
refactor: Improve the process of generating shots

sthnu32424 committed:
refactor: Constant 10 in the sense of strikes.

sthnu32424 committed:
refactor: Rename the test file

hor1shota commented:
配列は数値を変数に代入するとき、毎回初期化するようにしているのですが、これは冗長でしょうか？

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -1,45 +1,28 @@
-    shots << score.to_i
-  end
-end
+shots = scores.flat_map { |score| score == 'X' ? [STRIKE, 0] : score.to_i }
```
ここも工夫して短くしたんですね :+1:

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -1,45 +1,28 @@
 point = 0
-point += frames.each_with_index.sum do |frame, index|
-  # 10フレーム目はストライクやスペアを考慮せず合計を求める
+point = frames.each_with_index.sum do |frame, index|
```
一つ上の0での初期化なくていいかなと思いますよー

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,30 @@
+    frame.sum + frames[index + 1][0]
+  else
+    frame.sum
+  end
```
読みやすくなりましたね。ただこうなってくると、すべての分岐に`frame.sum`がいますよね。ということは分岐の外に出せるということですよね。するともっとスッキリかけると思うのでやってみてください :pray:

maedana replied:
これでは変数化しただけで、`frame_sum`がすべての分岐にあることは変わらないかなーと。ストライク、スペア、それ以外で3分岐してますけど、2分岐に出来るのでは?みたいな話です。

sthnu32424 committed:
refactor: change to use STRIKE where appropriate

https://github.com/hor1shota/ruby-practices/pull/3#discussion_r1952245627

sthnu32424 committed:
refactor: remove variable initialization

https://github.com/hor1shota/ruby-practices/pull/3#discussion_r1957476341

sthnu32424 committed:
refactor: change frame.sum call to once

https://github.com/hor1shota/ruby-practices/pull/3#discussion_r1957476982

sthnu32424 committed:
refactor: split methods to increase readability

hor1shota commented:
前回のSTRIKEの定数化が不適切な箇所で使用されていたため、改めて意味を考え、修正しました。
また、メソッド化を学習したので、可読性を上げるために実践してみました🙇‍♂️

sthnu32424 committed:
refactor: Make the call to frame.sum + only once

maedana commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,38 @@
+              frames[index + 1][0]
+            end
+
+    frame.sum + bonus.to_i
```
to_iは冗長ですかね。最初からIntegerではないですか?

sthnu32424 replied:
ストライクでもスペアでもない場合に nil が返ってしまい、演算子でエラーが発生し、
to_i を使うか、else 0 にするか迷いましたが、最終的に to_i を採用してエラーを回避することにしました。

maedana replied:
なるほどー。であれば`bonus`を0で初期化しておく、といいかしれないですね。`nil.to_i`は確かに0になりますけど、ちょっと意図が伝わりにくいコードになっていると思いますし。

sthnu32424 replied:
確かに`bonus.to_i`だと理解しづらいですね。
`else 0`にして、ストライク・スペア以外のフレームでは明示的に0を返すよう修正することにします！

hor1shota commented:
レビューありがとうございました！🙇‍♂️
